Voting on Proposals (#178)
magento/architecture: decision-making/voting-on-proposals.md

# Decision-making

### Proposal submission

> Proposal is submitted by opening pull request to this repository.

Author: PR contains proposal document (`.md` file).
Keeping it on GitHub allows to track document changes during the discussion thread.

> Initiator must present the idea on the architecture meeting.

Reviewer: Does the initiator have to wait for the next meeting or is this requirement independent of the 2 week timeline given below?

Author: I would say, initiator must present the idea and give 2 weeks discussion time. Need to work on wording.

Reviewer 2: Why does it need to be presented in a separate meeting? Or perhaps what types of changes need to go via this route and what kinf of changes can "just" go as PRs?

> The announcement should include problem description the proposal is trying to resolve as well as solution(s).

Reviewer 2: So does the proposal actually need to contain code? Or descriptions of the solution? A PR template might be helpful to not forget things

Reviewer 3: PR is a document or a set of documents. There is a template with suggested structure at https://github.com/magento/architecture/blob/master/design-documents/README.md

### Discussion

> There must be minimum 2 weeks of discussion in the pull request thread.

Reviewer: Some proposals are created from internal Magento stories or bugs with deadlines. Two weeks blocking a whole team or initiative is a long time. Are there exceptions to this rule? Maybe a more important question, are the guidelines clear for when a proposal MUST be made versus when a simple code review from an architect is sufficient?

Author: That's a really good point I would like to discuss. Maybe architecture team should initiate voting, not author.

Reviewer 4: Architecture team should initiate voting as for me, as some proposals may be misleading or propose the idea for which we already have resolution

After that initiator may decide when to start the vote.

### Voting
All eligible voters must get notification about the voting start.

> A valid voting period must be announced when voting is started, must be greater that 2 weeks and must not be changed during the vote.

Reviewer: Does this mean that every issue that is addressed from feedback resets the clock to 2 weeks?

Author: Hm, what I meant is when voting is started there should be no changes to the proposal.

> ### Approval

Reviewer: Are we ok with accepting proposals that have 100% opposition from Magento team but enough votes from community to qualify?

Author: From my point of view - yes, but I need more thoughts and concerns here.

Reviewer 5: More sophisticated model with someone having the power of veto may be implemented.

Reviewer 4: Don't think that we need to overcomplicate the voting process with veto right. I think every vote should have the same power.
For me, it's hard to imagine a situation when a proposal will get 100% opposition from Magento. It has never happened before, why should be foreseen in the future? 
Even internally for many questions, we don't have a single opinion.

Reviewer 6: *I fully support the initiative of public voting on proposals and will provide further input on design, goals, etc.* I am commenting on this facet separately for sake of expediency.

**A veto mechanism is neither necessary nor appropriate.** Whereas Magento, an Adobe company (MAAC) _alone_ releases the product we all call "Magento", I propose to add a note which recognizes the following reality:

> When a majority vote exists for a proposal, it does not guarantee that the proposal will be adopted or implemented.

This approach allows for two important realties:

1. To commemorate the collective opinion of the balloters in the face of disagreement with Magento/Adobe strategic needs
2. To allow Magento/Adobe to make even-better-informed product decisions based on our business needs.

As an addendum to (2), I believe this will force MAAC to be aware of & to justify product decisions which are in contrast to the wishes of the our internal & external ecosystem members, as proxied by balloters.

> The number of `approve` votes must be greater than or equal to the number of `reject` votes multiplied by two in order to merge pull request.

Reviewer 5: Maybe it's just me but wording seems complicated. Should we just stick to 2/3?

Reviewer 2: So the "shall we introduce this" vote needs 2/3 and possible "how shall we introduce this" need majority. so with 3 choices 1/3+1 might be sufficient..

In case of multi-choice voting (implementation details) the decision may be taken by simple plurality. This means that the voting option with the most votes wins. If there are multiple options with the most number of votes, the initiator can choose one of them.
Once pull request is merged the proposal is accepted.

# Eligible voters

> The group of eligible voters should be established. The group should represent the Magento community and include internal developers and architects (40%), Magento maintainers (30%), extension developers and partner contributors (30%).

Reviewer: Is there a requirement for how many votes are needed or is the number of votes required determined by the number of eligible people in the voter pool? Also, is there a requirement for percentage of votes from a particular category?

Author: Will add this info as well. This is the rough draft I wrote to understand if we want to move in this direction. It definitely requires defining all rules and use cases.

Reviewer 5: We might want to have some kind of rotation in "eligible voters" groups, otherwise there is a risk of forming a group biased to specific questions

Reviewer 4: I would not introduce tight coupling to the maintainers group establishing some proportion, because sometimes we have contributors who are not maintainers but actively participating in Magento Development process, there are other influencers who do not maintain Magento either.

This supposed to be a new role - Community Architect.

Magento team should be also represented widely, for example, there should be people from Cloud and MOM.

The proportion may vary, thus I would like not to mention it here at all, and we may extend the group of eligible voters with both internal and external participants. We should be flexible enough here.

Reviewer 2: Is there a minimum threshold of cast votes from the voters pool? so at least 50% of people eligible to vote need to vote otherwise the vote will... fail? need to be done again? will a failed vote need to be rewritten? or can it be voted on again immediately?

Reviewer 7: aka, what is quorum?

> The group of eligible voters should be established. The group should represent the Magento community and include internal developers and architects (40%), Magento maintainers (30%), extension developers and partner contributors (30%).

Reviewer 8: > The group should represent the Magento community and include internal developers and architects (40%)

We already have architects assigned to the components.

Author: Yes, it's a known fact. How do you propose to use those assignments?

Reviewer 8: Actually, it contradicts with the proposal: `The group should represent the Magento community and include internal developers and architects (40%)` because if all architects are included to the group, then the group will be too big otherwise the assigned architect might miss proposals for related components.

# Voting on Proposals

> Architecture team leads proposal handling and is responsible for procedure compliance.

Reviewer 9: ```suggestion
The Architecture team handles the proposal workflow and is responsible for procedural compliance.
```

> ### Proposal submission

Reviewer 9: ```suggestion
### Proposal submission workflow
```

> ### Proposal submission

Reviewer 9: ```suggestion
## Proposal submission
```

> - Proposal is submitted by opening pull request to this repository.

Reviewer 9: ```suggestion
- A proposal is submitted by opening a pull request to this repository.
```

> - Initiator must present the idea on the architecture meeting.

Reviewer 9: ```suggestion
- The initiator must present the idea in an architecture meeting.
```

> - The announcement should include a problem the proposal is trying to resolve as well as solution(s).

Reviewer 9: ```suggestion
- The announcement should include the problem the proposal is trying to resolve as well as any proposed solution(s).
```

> ### Discussion and Voting

Reviewer 9: ```suggestion
## Discussion and Voting
```

> - If there are controversial opinions in relation to the proposed idea architecture team may consider starting the voting.

Reviewer 9: ```suggestion
- If there are controversial opinions in relation to the proposed idea, the architecture team may consider starting the voting.
```

> - The voting is considered open when `voting` label is added to the pull request.

Reviewer 9: ```suggestion
- The voting is considered open when the `voting` label is added to the pull request.
```

- The voting can be represented as a single/multiple answer survey.

> - A valid voting period must be announced when the voting is started, must be greater that 1 week and must not be changed during the voting.

Reviewer 9: ```suggestion
- A valid voting period: 1) must be announced when the voting is started, 2) must be greater that 1 week and 3) must not be changed during the voting.
```

- The Discussion may continue in the pull request thread.

> - Voters cannot change their vote during the voting period.

Reviewer 3: Why not allow changing votes?

Author: My point was that it will lead to the carefully considered decision and more responsible attitude to the voting.

Reviewer: I've had my mind and vote changed many times after some good discussions. I think the idea you are trying to address is valid but I think there may be some validity to changing it as well

Reviewer 3: I'm not a psychologist, so maybe this is how people behave. So I'd just vote here to allow changing votes :) People may:
1. not notice that votes are not changeable until they really voted
1. click a wrong answer accidentally
1. vote and then see a discussion going in the PR (we still allow the discussion to go while voting is already in progress) and change their mind (thoughtfully)

Author: Makes sense. Updated this line.

> - Intermediate results should be hidden until the voting period ends.

Reviewer 9: ```suggestion
- Results should be hidden until the voting period ends.
```

> ### Results

Reviewer 9: ```suggestion
## Results
```

- Voting results should be added to the corresponding pull request.
- Participants should be mentioned by name.
- When a majority vote exists for a proposal, it does not guarantee that the proposal will be adopted or implemented.

> # Eligible voters

Reviewer 9: ```suggestion
## Eligible voters
```
